payments/models.py

```diff
@@ -154,3 +169,348 @@ def __str__(self):
             f"{self.service} - {self.period} "
             f"({self.price_value}{'€' if self.price_unit == PriceUnits.AMOUNT else '%'})"
         )
+
+
+class OrderManager(models.Manager):
+    def berth_orders(self):
+        return self.get_queryset().filter(
+            Q(
+                Q(_product_content_type__model=BerthProduct._meta.model_name)
+                | Q(_lease_content_type__model=BerthLease._meta.model_name)
+            )
+        )
+
+    def winter_storage_orders(self):
+        return self.get_queryset().filter(
+            Q(
+                Q(_product_content_type__model=WinterStorageProduct._meta.model_name)
+                | Q(_lease_content_type__model=WinterStorageLease._meta.model_name)
+            )
+        )
+
+
+class Order(UUIDModel, TimeStampedModel):
+    customer = models.ForeignKey(
+        "customers.CustomerProfile",
+        verbose_name=_("customer"),
+        on_delete=models.CASCADE,
+        related_name="orders",
+    )
+    product = GenericForeignKey("_product_content_type", "_product_object_id")
+    lease = GenericForeignKey("_lease_content_type", "_lease_object_id")
+    status = EnumField(OrderStatus, default=OrderStatus.WAITING)
+    comment = models.TextField(blank=True, null=True)
+    price = models.DecimalField(
+        verbose_name=_("price"),
+        max_digits=10,
+        decimal_places=2,
+        validators=[MinValueValidator(Decimal("0.01"))],
+        blank=True,
+    )
+    tax_percentage = models.DecimalField(
+        verbose_name=_("tax percentage"), max_digits=5, decimal_places=2, blank=True,
+    )
+    due_date = models.DateField(
+        verbose_name=_("due date"), default=calculate_order_due_date,
+    )
+
+    _product_content_type = models.ForeignKey(
+        ContentType,
+        on_delete=models.SET_NULL,
+        null=True,
+        blank=True,
+        related_name="product",
+    )
+    _product_object_id = models.UUIDField(
+        null=True, blank=True, verbose_name=_("product")
+    )
+
+    _lease_content_type = models.ForeignKey(
+        ContentType,
+        on_delete=models.SET_NULL,
+        null=True,
+        blank=True,
+        related_name="lease",
+    )
+    _lease_object_id = models.UUIDField(null=True, blank=True, verbose_name=_("lease"))
+
+    objects = OrderManager()
+
+    def __str__(self):
+        return f"{self.product} [{self.status}]"
+
+    @property
+    @rounded
+    def pretax_price(self):
+        return (
+            convert_aftertax_to_pretax(self.price, self.tax_percentage)
+            if self.price and self.tax_percentage
+            else None
```

Comment:
if it ever ends up in `None` it will crash `rounded` decorator, since it expects a decimal. should it ever be `we maybe drop the ip statement here?

Reply:
The reason to do it like that was (again) because of the admin. This time, the `OrderLine` inline on `Order` creation.

<img width="1224" alt="image" src="https://user-images.githubusercontent.com/15201480/84153144-2f9a7e80-aa6e-11ea-844d-b71e95a46589.png">

However, when adding that `else None`, everything goes fine.

<img width="1168" alt="image" src="https://user-images.githubusercontent.com/15201480/84153384-79836480-aa6e-11ea-8e60-581fed33d899.png">

Reply:
maybe then add the logic on the model admin, like:
```python
def pretax_price(self):
    return self.pretax_price or None
```
or something similar, just to not have that on the model directly?

Reply:
That would be ideal, but the error is raised on the division inside `convert_aftertax_to_pretax` when a`None` is passed instead of a number:
```
 return aftertax_price / (1 + tax_percentage / 100)
TypeError: unsupported operand type(s) for /: 'NoneType' and 'int'
```

(I tried doing that on the admin and didn't work either)

Reply:
Scratch that, if you check for `price/tax` on the admin, you can avoid the error. 👍